django/queryset/README.md

```diff
@@ -0,0 +1,340 @@
+# QuerySet
+
+## Introduction
+
+This guideline will base on these models:
+
+```python
+class User(models.Model):
+    username = models.CharField(max_length=50)
+
+
+class Person(models.Model):
+    user = models.ForeignKey(
+        'User',
+        on_delete=models.CASCADE,
+        related_name='people',
+    )
+    first_name = models.CharField(max_length=50)
+    last_name = models.CharField(max_length=150)
+    employee_id = models.CharField(max_length=15, unique=True)
+    active = models.Boolean(default=True, blank=True)
+
+
+class Department(models.Model):
+    name = models.CharField(max_length=255)
+
+    members = models.ManyToManyField(
+        'Person',
+        related_name='departments',
+    )
+```
+
+## Filtering
+
+### Joint Field Filtering
+
+```python
+# Yes
+def find_people_yes(query: str):
+    return Person.objects.annotate(
+        full_name=Concat('first_name', Value(' '), 'last_name'),
+    ).filter(
+        full_name__icontains=query,
+    )
+
+# No
+def find_people_no(query: str):
+    return Person.objects.filter(
+        Q(first_name__icontains=query)
+        | Q(last_name__icontains=query)
+    )
+```
+
+First, we try to filter people with name `John`...
+
+```python
+people = find_people_yes(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+
+people = find_people_no(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+```
+
+Everything seems fine, right? Both techniques return the same results. But what about filtering person's name by `John Doe`?
+
+```python
+people = find_people_yes(query='John Doe')
+# return: <QuerySet [<Person: John Doe>]>
+
+people = find_people_no(query='John Doe')
+# return: <QuerySet []>
+```
+
+Now you see the difference? It's because the second method (`find_people_no`) does not support filtering across two fields.
+
+So if you have a requirement like this, you should join two fields together before filter them.
+
+### Using `.count()` instead of `len()`
+
+```python
+# Yes
+Person.objects.count()     # SELECT COUNT(*) FROM person
+
+# No
+len(Person.objects.all())  # SELECT * FROM person
+```
+
+Selecting aggregate function from database always consume less time that getting every records.
```

Comment:
should change wording to

* calling `Person.objects.count()` will count directly inside DB.
* calling `len(Person.objects.all())` will fetch all data from DB and count in python.

So, using aggregate function from database always consume less time that getting every records.

Reply:
Updated krubb